test/interop/C/multilocale/globals/use_reliesOnGlobal.ml-test.c

```diff
@@ -1,3 +1,8 @@
+// This test currently fails sporadically. whatIsGlobal() reads a global that is
+// uninitialized. Globals are not default-value-initialized in multilocale runs
+// because they are just wide pointers to heap. See "library-init" version of
+// this test for the "correct" version of this.
```

Comment:
I don't think "correct" needs to be in quotation marks.  And I'd add that the correct version calls the library initialization function.

Reply:
Can/should we just remove this test since it has undefined behavior?

Reply:
I would prefer we not remove the test until it starts failing in a way that makes testing annoying.  I think it is important to know what happens when users make what is a pretty easy mistake to make, and if we can get something that will detect the error more deterministically, that would be ideal.

Reply:
Anything could happen though since it's UB. I think this is similar to having a test that has a double free or an OOB array access. We definitely have tests that do those things, but they run with runtime check enabled or are valgrind only or something like that. 

Could we make this valgrind only (maybe --no-local valgrind if --local valgrind doesn't fail because of heap promotion rules)?

Or if it's an easy mistake to make is there something we could do to help users, maybe have the compiler insert a check if module init ran before accessing a global (if boundsChecking is on)

Reply:
Sorry, lotta text incoming.

> Could we make this valgrind only (maybe --no-local valgrind if --local valgrind doesn't fail because of heap promotion rules)?

Valgrind doesn't work with multilocale runs, right?  Also, it would need to analyze a different program than what is running, since the place with the bad access is in the launched server rather than the client code.

> maybe have the compiler insert a check if module init ran before accessing a global (if boundsChecking is on)

I did consider that when first making the test, but I'm not sure where to insert it - before every access to a global variable?  The first access per function?  At the entrance of every function generally?  It seems like that'll get out of hand in large programs.

We'll also run into the problem that this is for a library that could be distributed more broadly - even if the user that wrote the library is good about calling the module initialization function, they can't guarantee that their users will be, so will always have to distribute a version with boundsChecking enabled, potentially hurting performance if they have a lot of global variables (which is reasonable for regular libraries).

It also is something that is completely unnecessary for normal Chapel code, so having these checks present means the user may get different behavior between using the .chpl file in a normal Chapel program or in a C or Python program.  This will make it difficult for library writers to test their program, since interop tests are harder to write than normal Chapel tests.  We really want to minimize the ways interop libraries are different from ordinary Chapel programs.

I've also considered ways to automatically call the module initialization function for users.  The initial solution I had for single locale libraries was bad and I haven't had the chance to revisit it yet.  It's potentially more possible for multilocale libraries than single locale libraries, but then you run into the problem of "well now the user has to make more changes between the programs used to test the multilocale and single locale compilations of the same library".

So we're kinda in a bind here.  There's a lot of competing values at play here, but I think the test fails rarely enough (and is run in limited enough settings) that I'd like to just leave it for now.

Reply:
I think the last justification for things being the way they are is that C programmers know more about memory issues - they're more likely to recognize a memory issue when they see it.  We don't run into this problem with Python because we linked the Chapel runtime init and the library init into a single call (and are working towards getting that called automatically iirc), and obviously don't have the same problem for Chapel programs because Chapel handles when the library is initialized itself.

Reply:
> Valgrind doesn't work with multilocale runs, right? Also, it would need to analyze a different program than what is running, since the place with the bad access is in the launched server rather than the client code.

Correct, that's why I was wondering if we could use --no-local with comm=none. When ASAN support comes online that may be a different option for multi-locale.

> I did consider that when first making the test, but I'm not sure where to insert it - before every access to a global variable? 

I would just have the compiler insert the check on every access to the global when in library mode. I don't think it's worth optimizing that check yet (we do bounds checks for every array access and that will have much more overhead than checking the global.)

I feel very uncomfortable intentionally leaving a test that has undefined behavior.


A program like:

```chpl
var a = c_malloc(int, 9);
a[9] = 9;
writeln(a[9]);
```

will appear to "work" on almost all platforms, but I don't think that makes it a valid test. I definitely understand wanting to try to test a likely user-error, but I really do not like intentionally testing UB.

My preference would be to remove or nontest this, and open an issue, but I won't continue to push for that if you're not convinced.

Reply:
I would much rather have a test that reminds us of the problem when it does fail than just an issue that gets buried and forgotten.  I'll open an issue and try to get to making the behavior consistent soon.